crates/sui-indexer-alt-jsonrpc/src/api/governance.rs

```diff
@@ -151,3 +67,96 @@ impl RpcModule for Governance {
         self.into_rpc()
     }
 }
+
+/// Load data and generate response for `getReferenceGasPrice`.
+async fn rgp_response(ctx: &Context) -> Result<BigInt<u64>, RpcError> {
+    use kv_epoch_starts::dsl as e;
+
+    let mut conn = ctx
+        .reader()
+        .connect()
+        .await
+        .context("Failed to connect to the database")?;
+
+    let rgp: i64 = conn
+        .first(
+            e::kv_epoch_starts
+                .select(e::reference_gas_price)
+                .order(e::epoch.desc()),
+        )
+        .await
+        .context("Failed to fetch the reference gas price")?;
+
+    Ok((rgp as u64).into())
+}
+
+/// Load data and generate response for `getLatestSuiSystemState`.
+async fn latest_sui_system_state_response(
+    ctx: &Context,
+) -> Result<SuiSystemStateSummary, RpcError> {
+    let wrapper: SuiSystemStateWrapper = fetch_latest(ctx, SUI_SYSTEM_STATE_OBJECT_ID)
+        .await
+        .internal_context("Failed to fetch system state wrapper object")?;
+
+    let inner_id = derive_dynamic_field_id(
+        SUI_SYSTEM_STATE_OBJECT_ID,
+        &TypeTag::U64,
+        &bcs::to_bytes(&wrapper.version).context("Failed to serialize system state version")?,
+    )
+    .context("Failed to derive inner system state field ID")?;
+
+    Ok(match wrapper.version {
+        1 => fetch_latest::<Field<u64, SuiSystemStateInnerV1>>(ctx, inner_id)
+            .await
+            .internal_context("Failed to fetch inner system state object")?
+            .value
+            .into_sui_system_state_summary(),
+        2 => fetch_latest::<Field<u64, SuiSystemStateInnerV2>>(ctx, inner_id)
+            .await
+            .internal_context("Failed to fetch inner system state object")?
+            .value
+            .into_sui_system_state_summary(),
+        v => rpc_bail!("Unexpected inner system state version: {v}"),
+    })
+}
+
+/// Fetch the latest version of the object at ID `object_id`, and deserialize its contents as a
+/// Rust type `T`, assuming that it is a Move object (not a package).
+///
+/// This function extracts the common parts of object loading for the latest system state object
+/// API, but it does not generalize beyond that, because it assumes that the objects being loaded
+/// are never deleted or wrapped and have always existed (because it loads using `LatestObjectKey`
+/// directly without checking the live object set).
+async fn fetch_latest<T: DeserializeOwned>(
+    ctx: &Context,
+    object_id: ObjectID,
+) -> Result<T, RpcError> {
+    let loader = ctx.loader();
+
+    let latest_version = loader
+        .load_one(LatestObjectKey(object_id))
+        .await
+        .context("Failed to load latest version")?
+        .ok_or_else(|| internal_error!("No latest version found"))?;
```

Comment:
Checking my understanding here: we use `internal_error` because we expect latest version to exist for sui system state inner. If this method was used for fetching latest version of a user provided object id, we would want to use `invalid_params` here. Is that right? If that's the case, maybe rename this function `fetch_latest_sui_system_state` to make it clear this behavior is only for fetching this object?

Reply:
Maybe I can call it `fetch_latest_for_system_state`, because it's used to fetch the system state object and also its dynamic field wrapper.

If we were trying to fetch the latest version of a user-provided ObjectID, we would first need to query `obj_info`, which would tell us if it exists or not, and only then would we query `obj_versions` to get its latest version, so it would still be an internal error for `obj_versions` to not produce a result -- it would mean that we pruned the version of a live object.